### PR TITLE
Bucket diagnostics

### PR DESCRIPTION
**Pull request by AlexisRenchon**
## Purpose 
This PR adds diagnostics to ClimaLand, as a new module in /src.
This module is built on ClimaDiagnostics.jl, which is also used by
ClimaAtmos. It contains default diagnostics methods for different
ClimaLand models (for now, BucketModel and SoilCanopyModel), those
diagnostics contains metadata such as where diagnostics variables are
stored in those models, what is there names (short, long, standard)
and physical units, with additional comments. It creates a Dict of those
variables, and also default Scheduling methods, such as hourly average
or monthly maximum. Users can define their own diagnostic variables or
scheduling, and developpers can add defaults. More information are
available in the documentation.

### CODE REVIEW

**Comment by AlexisRenchon** [comment]
@Sbozzolo 

Do you know why ClimaLandSimulations CI fail? 
The error is 
```
Run julia --project= -e 'using Pkg; Pkg.develop(path="$(pwd())"); Pkg.develop(path="$(pwd())/lib/ClimaLandSimulations")'
   Resolving package versions...
ERROR: Unsatisfiable requirements detected for package ClimaDiagnostics [1ecacbb8]:
 ClimaDiagnostics [1ecacbb8] log:
 ├─possible versions are: 0.0.1-0.0.5 or uninstalled
 └─restricted to versions 0.2 by ClimaLand [08f[4](https://github.com/CliMA/ClimaLand.jl/actions/runs/9321596680/job/25660930914?pr=628#step:5:5)d4ce] — no versions left
   └─ClimaLand [08f4d4ce] log:
     ├─possible versions are: 0.12.2 or uninstalled
     └─ClimaLand [08f4d4ce] is fixed to version 0.12.2
 ```
But if I run 
```
julia --project= -e 'using Pkg; Pkg.develop(path="$(pwd())"); Pkg.develop(path="$(pwd())/lib/ClimaLandSimulations")'
```
locally, it works just fine

**Comment by Sbozzolo** [comment]
Maybe you have to add ClimaDiagnostics to the Project.toml for ClimaLandSimulations?

File: src/Diagnostics/standard_diagnostic_frequencies.jl
Comment (Sbozzolo) [inline src/Diagnostics/standard_diagnostic_frequencies.jl:239]:
```suggestion
"""
    hourly_mins(short_names...; output_writer, t_start)

Return a `ScheduledDiagnostics` that computes the hourly min for the given variable.
"""

```

Comment (Sbozzolo) [reply src/Diagnostics/standard_diagnostic_frequencies.jl:239]: (Remove one extra line from the docstring)

File: src/Diagnostics/diagnostic.jl
Comment (Sbozzolo) [inline src/Diagnostics/diagnostic.jl:70]:
```suggestion

```

File: src/Diagnostics/bucket_compute_methods.jl
Comment (Sbozzolo) [inline src/Diagnostics/bucket_compute_methods.jl:9]: We can try coming up with a macro to write this

Comment (Sbozzolo) [reply src/Diagnostics/bucket_compute_methods.jl:9]: Might look like:

```julia
macro generate_error_functions(function_names...)
    functions = Expr[]
    for function_name in function_names
        function_name_sym = Symbol("compute_", function_name, "!")
        body = quote
            function $function_name_sym(_, _, _, _, land_model)
                error_diagnostic_variable($(Symbol(function_name)), land_model)
            end
        end
        push!(functions, body)
    end
    return quote
        $(functions...)
    end
end

@generate_error_functions "net_radiation" "surface_temperature" "surface_specific_humidity"
```

File: src/Diagnostics/default_diagnostics.jl
Comment (Sbozzolo) [inline src/Diagnostics/default_diagnostics.jl:13]:
```suggestion
    common_diagnostics(period, reduction)
```

File: docs/src/diagnostics/developpers_diagnostics.md
Comment (Sbozzolo) [inline docs/src/diagnostics/developpers_diagnostics.md:1]: developer is spelled with only one `p`

File: docs/src/diagnostics/users_diagnostics.md
Comment (Sbozzolo) [inline docs/src/diagnostics/users_diagnostics.md:1]: This page has to include instruction how to add a diagnostic that is not default

**Comment by AlexisRenchon** [comment]
@kmdeck Could you just look at `src/Diagnostics/define_diagnostics.jl` and check if the names, units, and comments look good? No need to review the full PR! 

Thank you!!

**Comment by AlexisRenchon** [comment]
@Sbozzolo 

- I moved the experiment to test (there is nothing new in experiment now). The idea is having a global bucket model in test that then calls some Diagnostic functions and test them. To do that, I had to add SciMLBase and ClimaTimeSteppers to the test Project.toml, I am not sure if this is something we are trying to avoid. 
- I added comments and changed some names to `src/Diagnostics/define_diagnostics.jl`
- I added your macro to generate error functions 
- Docs: available_diagnostics works well
- Docs: I added more content to developers and users documentation page, but there is still more to add (and add more links)
- All CI checks passed

File: Artifacts.toml
Comment (Sbozzolo) [inline Artifacts.toml:5]: This should probably not be here

File: docs/src/diagnostics/make_diagnostic_table.jl
Comment (Sbozzolo) [inline docs/src/diagnostics/make_diagnostic_table.jl:5]:
```suggestion
CL.define_diagnostics!(nothing)
```

File: src/Diagnostics/diagnostic.jl
Comment (Sbozzolo) [inline src/Diagnostics/diagnostic.jl:88]: For your reference, as discussed
```julia
@test_throws ErrorException get_diagnostic_variable("bob")
```

File: src/Diagnostics/diagnostic.jl
Comment (Sbozzolo) [inline src/Diagnostics/diagnostic.jl:84]:
```suggestion

```

File: src/Diagnostics/bucket_compute_methods.jl
Comment (Sbozzolo) [inline src/Diagnostics/bucket_compute_methods.jl:22]: To test this
```julia
@test_throws ErrorException compute_albedo!(_, _, _, _, Nothing)
```

File: docs/src/diagnostics/users_diagnostics.md
Comment (Sbozzolo) [inline docs/src/diagnostics/users_diagnostics.md:73]: What if I want to compute some existing diagnostics but with different Writers? Or with different frequencies?

File: src/Diagnostics/define_diagnostics.jl
Comment (kmdeck) [inline src/Diagnostics/define_diagnostics.jl:26]: "Difference" instead of balance?. Maybe we can also note that a positive value means radiation is absorbed by land?

File: src/Diagnostics/define_diagnostics.jl
Comment (kmdeck) [inline src/Diagnostics/define_diagnostics.jl:37]: How is T_sfc defined? In the bucket, I think we use the value at the top center. that might be better described as "Temperature of the land surface". But if this is defined as a skin temperature (or something), what you have is OK.

File: src/Diagnostics/define_diagnostics.jl
Comment (kmdeck) [inline src/Diagnostics/define_diagnostics.jl:59]: maybe "due to evaporation/sublimation"

File: src/Diagnostics/define_diagnostics.jl
Comment (kmdeck) [inline src/Diagnostics/define_diagnostics.jl:89]: "Water vapor flux"? since this could be sublimation too

Comment (kmdeck) [reply src/Diagnostics/define_diagnostics.jl:89]: also as a comment - this is converted from a mass flux into an equivalent volume flux of liquid water. maybe this is what you meant

File: src/Diagnostics/define_diagnostics.jl
Comment (kmdeck) [inline src/Diagnostics/define_diagnostics.jl:103]: and -> at

File: src/Diagnostics/define_diagnostics.jl
Comment (kmdeck) [inline src/Diagnostics/define_diagnostics.jl:127]: W here is a scalar, not at each layer

**Review by kmdeck — COMMENTED** [review]
Looks good in src/Diagnostics/define_diagnostics.jl! Thanks!

File: src/Diagnostics/Diagnostics.jl
Comment (kmdeck) [inline src/Diagnostics/Diagnostics.jl:7]: ah ha!

File: .buildkite/pipeline.yml
Comment (Sbozzolo) [inline .buildkite/pipeline.yml:144]:
```suggestion
        artifact_paths: "experiments/standalone/Bucket/artifacts_temporalmap/*cpu*"
```

File: src/Diagnostics/Diagnostics.jl
Comment (Sbozzolo) [reply src/Diagnostics/Diagnostics.jl:7]: The problem was that `Diagnostics.jl`  was imported before the soil module

File: src/Diagnostics/default_diagnostics.jl
Comment (Sbozzolo) [inline src/Diagnostics/default_diagnostics.jl:113]: Add new line here

File: src/Diagnostics/define_diagnostics.jl
Comment (Sbozzolo) [inline src/Diagnostics/define_diagnostics.jl:431]: If comments are useless, let's not put them, or let's write `TODO` instead.

File: src/Diagnostics/soilcanopy_compute_methods.jl
Comment (Sbozzolo) [inline src/Diagnostics/soilcanopy_compute_methods.jl:7]:
```suggestion
        out .= p.soil.R_n
```

File: src/Diagnostics/soilcanopy_compute_methods.jl
Comment (Sbozzolo) [inline src/Diagnostics/soilcanopy_compute_methods.jl:31]: Is this `p.bucket`?

File: src/Diagnostics/define_diagnostics.jl
Comment (AlexisRenchon) [reply src/Diagnostics/define_diagnostics.jl:26]: I am not sure where, but we should say somewhere that our sign convention throughout is relative to atmosphere or land, e.g., positive if atmosphere gain, negative if atmosphere loss... (is our sign actually consistent for all? e.g., negative values for rain?)

File: src/Diagnostics/define_diagnostics.jl
Comment (AlexisRenchon) [reply src/Diagnostics/define_diagnostics.jl:37]: I will leave this comment up for now. We should decide soon. What do you think @braghiere ?

File: src/Diagnostics/define_diagnostics.jl
Comment (AlexisRenchon) [reply src/Diagnostics/define_diagnostics.jl:89]: `"Flux of water from the land surface to the atmosphere. E.g., evaporation or sublimation."`
Not sure if I should be e.g., or i.e., I guess there's no transpiration in bucket, so i.e., may be ok (if these are the only 2 options)

the units are `"m s^-1"`, correct? I guess this implies this is a volume of water so no need to specify in comments?

File: src/Diagnostics/define_diagnostics.jl
Comment (braghiere) [reply src/Diagnostics/define_diagnostics.jl:37]: I agree with @kmdeck. Usually surface temperature refers to skin temperature (a combination of soil + canopy). Temperature at the land-atmosphere interface sounds more like skin temperature to me. How is this term calculated? What does this actually physically represent?

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:175]: Let's use a more descriptive name, maybe `global_bucket_function`? 

This will help when we add the diagnostics to more experiments

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:183]: This 1.0 is unclear

File: src/Diagnostics/bucket_compute_methods.jl
Comment (AlexisRenchon) [reply src/Diagnostics/bucket_compute_methods.jl:22]:
```
julia> using Test

julia> using ClimaLand

julia> @test isdefined(ClimaLand.Diagnostics, :compute_albedo!)
Test Passed

julia> @test_throws ErrorException("cannot albedo with nothing") ClimaLand.Diagnostics.comput
e_albedo!(_, _, _, _, nothing)
ERROR: syntax: all-underscore identifier used as rvalue around REPL[9]:1
Stacktrace:
 [1] top-level scope
   @ REPL[9]:1
```

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:194]:
```suggestion
simdir = ClimaAnalysis.SimDir(output_dir)
```

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:200]:
```suggestion
CairoMakie.save(joinpath(output_dir, "alpha.png"), fig)
```

File: src/Diagnostics/diagnostic.jl
Comment (AlexisRenchon) [reply src/Diagnostics/diagnostic.jl:88]: I added that test, BUT, currently we do not call `define_diagnostics!` inside src (only in docs), so `get_diagnostic_variable` would throw an error with any variable...

File: docs/src/diagnostics/make_diagnostic_table.jl
Comment (AlexisRenchon) [reply docs/src/diagnostics/make_diagnostic_table.jl:5]: I added that, but while this works, we can't use `get_diagnostic_variable` when just using ClimaLand.Diagnostics since we don't call `define_diagnostics` in src

File: src/Diagnostics/bucket_compute_methods.jl
Comment (AlexisRenchon) [reply src/Diagnostics/bucket_compute_methods.jl:5]: the test didn't work, let's discuss this when we meet

File: src/Diagnostics/diagnostic.jl
Comment (Sbozzolo) [reply src/Diagnostics/diagnostic.jl:88]: You can call define_diagnostics in your test

File: .github/workflows/ClimaLandSimulations.yml
Comment (AlexisRenchon) [reply .github/workflows/ClimaLandSimulations.yml:22]: Let's go back to this at some point, as discussed, this still doesn't work

File: src/Diagnostics/define_diagnostics.jl
Comment (AlexisRenchon) [reply src/Diagnostics/define_diagnostics.jl]: I believe this is the temperature at the center of the top layer, as Kat explained, soil layers being discretized to n layers, n defined by users. We have T_sfc stored for the bucket, but no T_sfc for soilcanopymodel, at the moment, I believe (maybe this has changed). We could rename T_sfc to T_skin. (we have T at all soil layers for both models, but T_sfc only for bucket)

We should meet, the 3 of us, to discuss ClimaLand diagnostic variables. 
Sometimes, we compute the same things across models (bucket, soilcanopymodel, other), they will be stored differently (e.g., p.soil.T or Y.bucket.T), but will have the same name, units, comments.

Before we merge this PR, it will be good to go through the list of variables, and make sure we agree on all the naming (short long standard), units, and comments. I will have something for all, but we should make sure we all agree and it looks good together.

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (AlexisRenchon) [reply experiments/standalone/Bucket/global_bucket_function.jl]: I agree. I just put a value for `t_start`, but maybe we don't need it? It's an optional arg in `EveryDtSchedule(period; t_start)` called in `common_diagnostics`

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:176]:
```suggestion
base_output_dir = "global_bucket_function/"
output_dir = ClimaUtilities.OutputPathGenerator.generate_output_path(base_output_dir)
```

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [reply experiments/standalone/Bucket/global_bucket_function.jl]: Yes, `t_start` is good!

File: src/Diagnostics/define_diagnostics.jl
Comment (braghiere) [reply src/Diagnostics/define_diagnostics.jl]: Sounds good to me.

File: .github/workflows/ClimaLandSimulations.yml
Comment (Sbozzolo) [reply .github/workflows/ClimaLandSimulations.yml:22]: Maybe leave a comment on why this is commented out

File: docs/make.jl
Comment (Sbozzolo) [inline docs/make.jl:10]: Put this back

File: docs/src/diagnostics/users_diagnostics.md
Comment (Sbozzolo) [inline docs/src/diagnostics/users_diagnostics.md:23]: Update this with the OuputPathGenerator

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:172]: This
`saving_cb = ClimaLand.NonInterpSavingCallback(saved_values, saveat);` is no longer required if we use diagnostics

Comment (Sbozzolo) [reply experiments/standalone/Bucket/global_bucket_function.jl:172]: From L161 to L168

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:197]:
```suggestion
```

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (Sbozzolo) [inline experiments/standalone/Bucket/global_bucket_function.jl:200]:
```suggestion
for short_name in ["alpha", "W"]
var = get(simdir; short_name)
fig = CairoMakie.Figure(size = (400, 600))
viz.plot!(fig, alpha, time = 2 * 86400)
CairoMakie.save(joinpath(output_dir, "$short_name.png"), fig)
end
```

File: lib/ClimaLandSimulations/Project.toml
Comment (Sbozzolo) [inline lib/ClimaLandSimulations/Project.toml:49]: Double check if you need this

File: src/ClimaLand.jl
Comment (Sbozzolo) [inline src/ClimaLand.jl:325]:
```suggestion
```

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (AlexisRenchon) [reply experiments/standalone/Bucket/global_bucket_function.jl]: Done and it works, but two issues: 

- tsoil is 3D (has depth) so viz.plot! requires z. But other variables don't have z. For now I just commented out tsoil, but this is an easy fix.
- more importantly, in this experiment it seems all variables are constant in space, so the heatmap just shows one color.

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (AlexisRenchon) [reply experiments/standalone/Bucket/global_bucket_function.jl:172]: Let's talk about this, 
```
W = [
    Array(Remapping.interpolate(remapper, sol.u[k].bucket.W)) for
    k in 1:length(sol.t)
];
```
returns a `9-element Vector{Matrix{Float64}}`

would be equivalent to ( ? )
```
W = get(simdir; short_name = "wsoil")
W = W.data
```
which returns `23×180×90 Array{Float64, 3}`

(do we need to adjust format a bit, or am I getting the wrong this?)

File: experiments/standalone/Bucket/global_bucket_function.jl
Comment (AlexisRenchon) [reply experiments/standalone/Bucket/global_bucket_function.jl]: Let's come back to this too - I am not sure I followed, do we keep the 1.0? Or adjust something?

**Comment by AlexisRenchon** [comment]
Let's discuss the Bucket experiment a bit, we want to use an existing experiment, but it would be good if variables vary in space at least... I am not sure it is the case with `global_bucket_function.jl` (our heatmap plots are 1 color), maybe we should use another. We were using `global_bucket_temporalmap.jl` because of this, which has albedo varying in space (read from a file), but it defines the bucket model inside a function that I commented out before... and we want to not change the previous code but add diagnostics to it. Anyway, let's discuss.

File: lib/ClimaLandSimulations/Project.toml
Comment (AlexisRenchon) [reply lib/ClimaLandSimulations/Project.toml:49]: I think we should comeback to `ClimaLandSimulations` in another PR, it has a lot of things to adjust...

**Comment by Sbozzolo** [comment]
Let's not forget about the issue on `@generate_error_functions`

**Review by Sbozzolo — APPROVED** [review]
Some work is left for a future PR, but this looks good!

Thank you very much!

**Pull request merged**
